hardware/eVehicles/MercApi.cpp

```diff
@@ -0,0 +1,845 @@
+}
+
+bool CMercApi::GetVehicleData(tVehicleData& data)
+{
+	Json::Value reply;
+	bool bData = false;
+
+	if (GetData("vehiclelockstatus", reply))
+	{
+		if (reply.size() == 0)
+		{
+			bData = true;	// This occurs when the API call return a 204 (No Content). So everything is valid/ok, just no data
+		}
+		else
+		{
+			if (!reply.isArray())
+			{
+				_log.Log(LOG_ERROR, "MercApi: Unexpected reply from VehicleLockStatus.");
+			}
+			else
+			{
+				GetVehicleData(reply, data);
+				bData = true;
+			}
+		}
+	}
+
+	reply.clear();
+
+	if (GetData("payasyoudrive", reply))
+	{
+		if (reply.size() == 0)
+		{
+			bData = true;	// This occurs when the API call return a 204 (No Content). So everything is valid/ok, just no data
+		}
+		else
+		{
+			if (!reply.isArray())
+			{
+				_log.Log(LOG_ERROR, "MercApi: Unexpected reply from PayasyouDrive.");
+			}
+			else
+			{
+				GetVehicleData(reply, data);
+				bData = true;
+			}
+		}
+	}
+
+	return bData;
+}
+
+bool CMercApi::GetCustomData(tCustomData& data)
+{
+	Json::Value reply;
+	bool bCustom = true;
+
+	if (m_capabilities.has_custom_data)
+	{
+		std::vector<std::string> strarray;
+
+		StringSplit(m_fields, ",", strarray);
+		for(uint8_t i=0; i<strarray.size(); i++)
+		{
+			reply.clear();
+			if (GetResourceData(strarray[i], reply))
+			{
+				if(reply.size() == 0)
+				{
+					_log.Debug(DEBUG_NORM, "MercApi: Got empty data for resource %s", strarray[i].c_str());
+				}
+				else
+				{
+					//_log.Debug(DEBUG_NORM, "MercApi: Got data for resource %s :\n%s", strarray[i].c_str(),reply.toStyledString().c_str());
+
+					if (!reply[strarray[i]].empty())
+					{
+						if (reply[strarray[i]].isMember("value"))
+						{
+							std::string resourceValue = reply[strarray[i]]["value"].asString();
+
+							Json::Value customItem;
+							customItem["id"] = i;
+							customItem["value"] = resourceValue;
+							customItem["label"] = strarray[i];
+
+							data.customdata.append(customItem);
+
+							_log.Debug(DEBUG_NORM, "MercApi: Got data for resource (%d) %s : %s", i, strarray[i].c_str(), resourceValue.c_str());
+						}
+					}
+				}
+			}
+			else
+			{
+				_log.Debug(DEBUG_NORM,"MercApi: Failed to retrieve data for resource %s!", strarray[i].c_str());
+			}
+		}
+	}
+
+	return bCustom;
+}
+
+void CMercApi::GetVehicleData(Json::Value& jsondata, tVehicleData& data)
+{
+	uint8_t cnt = 0;
+	do
+	{
+		Json::Value iter;
+
+		iter = jsondata[cnt];
+		for (auto const& id : iter.getMemberNames())
+		{
+			if(!(iter[id].isNull()))
+			{
+				_log.Debug(DEBUG_NORM, "MercApi: Found non empty field %s", id.c_str());
+
+				if (id == "doorlockstatusvehicle")
+				{
+					Json::Value iter2;
+					iter2 = iter[id];
+					if(!iter2["value"].empty())
+					{
+						_log.Debug(DEBUG_NORM, "MercApi: DoorLockStatusVehicle has value %s", iter2["value"].asString().c_str());
+						data.car_open = (iter2["value"].asString() == "1" || iter2["value"].asString() == "2" ? false : true);
```

Comment:
What if a window is open? The tesla api considers everything that can be used to access the vehicle as an unlocked status (even though the car is 'locked', the window might be open hence the car is not locked). I would check all possible openings here ;-)

Reply:
Yes, I was thinking about that too. At least adjust the message if something is open. I look into it.